src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/Helpers.cs

```diff
@@ -127,12 +130,31 @@ internal static bool TextTypedSoFarMatchesItem(RoslynCompletionItem item, string
         }
 
         // Tab, Enter and Null (call invoke commit) are always commit characters. 
-        internal static bool IsStandardCommitCharacter(char c)
+        public static bool IsStandardCommitCharacter(char c)
             => c is '\t' or '\n' or '\0';
 
         // This is a temporarily method to support preference of IntelliCode items comparing to non-IntelliCode items.
         // We expect that Editor will introduce this support and we will get rid of relying on the "★" then.
-        internal static bool IsPreferredItem(this VSCompletionItem completionItem)
+        public static bool IsPreferredItem(this VSCompletionItem completionItem)
             => completionItem.DisplayText.StartsWith("★");
+
+        public static ImmutableArray<EditorAsyncCompletionData.CompletionFilterWithState> CombineFilterStates(
+            ImmutableArray<EditorAsyncCompletionData.CompletionFilterWithState> states1, ImmutableArray<EditorAsyncCompletionData.CompletionFilterWithState> states2)
+        {
+            var filterStateMap = PooledDictionary<EditorAsyncCompletionData.CompletionFilter, bool>.GetInstance();
+            AddFilterState(states1);
+            AddFilterState(states2);
```

Comment:
is it ok that the arrays might get reordered because dictionaries don't guarantee order?

Reply:
That's a good point. I'm not 100% sure if it's intended, but it's indeed guaranteed by how `FilterSet` is implemented now. I think the ordering makes sense so I will keep it that way with merge.